### PR TITLE
[Model] RowParallelLinear: pass bias to quant_method.apply 

### PR DESCRIPTION
**Pull request by tdoublep**
I think for TP=1 case we can pass the `bias` to the apply function, which should lead to more use of the fused CUTLASS kernels.

cc @tlrmchlsmth @cyang49 

**Update**: this is now implemented for TP>1 too

### CODE REVIEW

**Comment by tdoublep** [comment]
CI failures are a bit weird (see https://github.com/vllm-project/vllm/issues/6332)

File: vllm/model_executor/layers/linear.py [hunk 1]
```diff
@@ -753,18 +753,23 @@ def forward(self, input_):
 
         # Matrix multiply.
         assert self.quant_method is not None
-        output_parallel = self.quant_method.apply(self, input_parallel)
+        bias_ = None if (self.tp_size > 1 or self.skip_bias_add) else self.bias
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/linear.py:756]: Could you add a comment explaining what happens when tp_size > 1? I think it should say something like:
```
# Fuse bias into the GEMM in the TP == 1 case.
# The TP > 1 case is problematic as the bias will be redundantly summed during the AllReduce if fused
```

**Review by tlrmchlsmth — COMMENTED** [review]
This looks good. I see that for the TP > 1 case we can't naively fuse the bias, since each rank will add the bias to its output and then the biases will be redundantly summed during the AllReduce. 

Another thing to try is to pick one of the ranks to apply the bias. The overhead of applying a bias during a GEMM is low enough that we probably don't need to worry load imbalance, and this will let every rank skip a load and store of the activation tensor.

> which should lead to more use of the fused CUTLASS kernels.

FYI this won't lead to more use of the CUTLASS kernels. Currently they're used whenever there is _no_ bias. The issue

**Comment by tlrmchlsmth** [comment]
>CI failures are a bit weird (see https://github.com/vllm-project/vllm/issues/6332) 

If the bias is fused into the GEMM, it will likely be added to the output at a higher precision. For example in our CUTLASS fp8 kernels, the bias add happens in fp32.  On main the output will be converted to the output dtype before adding the bias. That could explain the test failure you're seeing.

**Comment by tdoublep** [comment]
> For example in our CUTLASS fp8 kernels, the bias add happens in fp32. On main the output will be converted to the output dtype before adding the bias. That could explain the test failure you're seeing.

Thanks, yes that explains why this change causes the tests to fail (since the same tests also fail in fp32). 

Now need to figure out why tests don't work in fp32, it is somehow related to chunked prefill tests only.

**Comment by tdoublep** [comment]
OK, so the tests on this branch are now passing in fp32 but only after this unrelated bug is fixed: https://github.com/vllm-project/vllm/pull/6373

> On main the output will be converted to the output dtype before adding the bias

Is this the desired behaviour? Is there branch where this doesn't happen that I can test against?

**Comment by tdoublep** [comment]
The offending tests actually also fail on `main` using float16 on an H100. I think probably the test in question (`tests/samplers/test_logprobs.py::test_get_prompt_logprobs`) should be changed to use float32 in order to get a reliable comparison against the logprobs from HF.

File: vllm/model_executor/layers/linear.py (hunk 1 above)
Comment (tdoublep) [reply vllm/model_executor/layers/linear.py]: Added a comment (although it reads a bit differently since the TP>1 case is now handled).

**Comment by tdoublep** [comment]
>Another thing to try is to pick one of the ranks to apply the bias. The overhead of applying a bias during a GEMM is low enough that we probably don't need to worry load imbalance, and this will let every rank skip a load and store of the activation tensor.

I like that idea - I implemented it and it makes the code even simpler. 

> FYI this won't lead to more use of the CUTLASS kernels. Currently they're used whenever there is no bias. 

Right, but they will be after we merge #6270 

> The offending tests actually also fail on main using float16 on an H100. 

The CI tests should pass after we merge #6409

**Review by tlrmchlsmth — COMMENTED** [review]
This looks good to me, although the spec decoding tests look suspicious, so I'll wait until those are resolved before accepting

Have you done any e2e benchmarking of e.g. Granite or Qwen?

**Comment by tdoublep** [comment]
> This looks good to me, although the spec decoding tests look suspicious
I agree - will look into it.

> Have you done any e2e benchmarking of e.g. Granite or Qwen?
Not yet but tomorrow hopefully, will share once I have some data.

**Comment by tdoublep** [comment]
@tlrmchlsmth I fixed the CI issue by changing the failing TP=2 test to use `float32` precision (which we are already using for the [equivalent TP=1 test](https://github.com/vllm-project/vllm/blob/main/tests/spec_decode/e2e/test_mlp_correctness.py#L37)).

**Review by tlrmchlsmth — APPROVED** [review]
LGTM

**Comment by tlrmchlsmth** [comment]
@tdoublep could you share perf results when you have them? Curious about what the impact will be. Thanks!

**Comment by tdoublep** [comment]
@tlrmchlsmth got sidetracked with a few things, but had some time to run benchmarks today. I ran the benchmarking using `granite-20b` using an FP8 checkpoint I created myself using AutoFP8. I'm using our [fmperf](https://github.com/fmperf-project/fmperf) tool and I simulate different number of concurrent users of the server (indicated as the point labels on the plots below), each user is sending requests with 128 input tokens, 128 output tokens. I've tested 3 variants:
1. `main` branch but with the changes from #6270 reverted (e.g., so without fusion of bias adds)
2. `main` branch unchanged
3. `rplinear_bias` branch, which should have *more* fused bias adds

Here are the results:

image

The results look as expected imo

cc @cyang49

**Comment by tlrmchlsmth** [comment]
The results look great!

>main branch but with the changes from https://github.com/vllm-project/vllm/pull/6270 reverted (e.g., so without fusion of bias adds)

Just one detail: If you revert #6270, the effect isn't that the bias adds aren't fused. Instead we were falling back to `torch._scaled_mm`

Thanks for linking `fmperf`, I'll check that out

**Review by mgoin — APPROVED** [review]
Nice simplicity with the final version

**Comment by tdoublep** [comment]
> Just one detail: If you revert https://github.com/vllm-project/vllm/pull/6270, the effect isn't that the bias adds aren't fused. Instead we were falling back to torch._scaled_mm

You are right - have updated the comment + figure accordingly.

**Pull request merged**